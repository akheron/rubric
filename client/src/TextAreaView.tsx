import React, { useState } from 'react'
import { SelectionType } from './types'
import DOMPurify from 'dompurify'

interface Props {
  title: string
  selection: SelectionType
}

const TextAreaView = (props: Props): JSX.Element => {
  const [value, setValue] = useState('')

  const handleValueChange = (
    event: React.ChangeEvent<HTMLTextAreaElement>
  ): void => {
    setValue(event.target.value)
  }

  const escapeHtml = (unsafe: string) =>
    unsafe
      .replace(/&/g, '&amp;')
      .replace(/</g, '&lt;')
      .replace(/>/g, '&gt;')
      .replace(/"/g, '&quot;')
      .replace(/'/g, '&#039;')

  const getSanitizedHTMLValue = (value: string) =>
    DOMPurify.sanitize(escapeHtml(value).replaceAll(/\n/g, '<br />'))

  const valueComponent =
    props.selection === 'select' ? (
      <span
        dangerouslySetInnerHTML={{
          __html: getSanitizedHTMLValue(value),
        }}
      />
    ) : (
      <textarea rows={4} cols={80} value={value} onChange={handleValueChange} />
    )

  return (
    <table className="textAreaCriterion">
      <tbody>
        <tr>
          <td className="textAreaCriterionCell">
            <div className="criterionTitle">{props.title}</div>
            <div>{valueComponent}</div>
          </td>
        </tr>
      </tbody>
    </table>
  )
}

export default TextAreaView
