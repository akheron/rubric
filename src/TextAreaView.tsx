import React, { useState } from 'react'

interface Props {
  title: string
  selected: boolean
}

const TextAreaView = (props: Props): JSX.Element => {
  const [value, setValue] = useState('')

  const handleValueChange = (
    event: React.ChangeEvent<HTMLTextAreaElement>
  ): void => {
    setValue(event.target.value)
  }

  const valueComponent = props.selected ? (
    <span>{value}</span>
  ) : (
    <textarea
      rows={4}
      cols={80}
      value={value}
      onChange={handleValueChange}
    ></textarea>
  )

  return (
    <table className="textAreaCriterion">
      <tr>
        <td className="textAreaCriterionCell">
          <div className="criterionTitle">{props.title}</div>
          <div>{valueComponent}</div>
        </td>
      </tr>
    </table>
  )
}

export default TextAreaView
