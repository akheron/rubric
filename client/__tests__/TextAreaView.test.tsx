import React from 'react'
import renderer from 'react-test-renderer'
import TextAreaView from '../src/TextAreaView'
import DOMPurify from 'dompurify'

describe('TextAreaView renders correctly with', () => {
  const expectValidComponent = (
    title: string,
    value: string,
    selection = ''
  ): void => {
    const component = renderer.create(
      <TextAreaView title={title} value={value} selection={selection} />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  }

  describe('no selection and', () => {
    it('empty title and empty value', () => {
      expectValidComponent('', '')
    })
    it('text title and empty value', () => {
      expectValidComponent('xyzzy', '')
    })
    it('text title and text value', () => {
      expectValidComponent('foo', 'bar')
    })
    it('text title and value with newlines', () => {
      expectValidComponent('foo', 'bar\n\nbaz')
    })
    it('text title and value with XSS attack', () => {
      expectValidComponent(
        'title',
        '<span onmouseover="alert(document.cookie)">this is an\nXSS attack</span>'
      )
    })
  })

  describe('selection and', () => {
    it('empty title and empty value', () => {
      expectValidComponent('', '', 'select')
    })
    it('text title and empty value', () => {
      expectValidComponent('xyzzy', '', 'select')
    })
    it('text title and text value', () => {
      expectValidComponent('foo', 'bar', 'select')
    })
    it('text title and value with newlines', () => {
      expectValidComponent('foo', 'bar\n\nbaz', 'select')
    })
    it('text title and value with XSS attack', () => {
      expectValidComponent(
        'title',
        '<span onmouseover="alert(document.cookie)">this is an\nXSS attack</span>',
        'select'
      )
    })
  })
})
