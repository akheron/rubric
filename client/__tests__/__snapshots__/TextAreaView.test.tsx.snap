// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAreaView renders correctly when no selection and empty title and empty value 1`] = `
<table
  className="textAreaCriterion"
>
  <tbody>
    <tr>
      <td
        className="textAreaCriterionCell"
      >
        <div
          className="criterionTitle"
        />
        <div>
          <textarea
            cols={80}
            onChange={[Function]}
            rows={4}
            value=""
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TextAreaView renders correctly when no selection and text title and empty value 1`] = `
<table
  className="textAreaCriterion"
>
  <tbody>
    <tr>
      <td
        className="textAreaCriterionCell"
      >
        <div
          className="criterionTitle"
        >
          xyzzy
        </div>
        <div>
          <textarea
            cols={80}
            onChange={[Function]}
            rows={4}
            value=""
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TextAreaView renders correctly when no selection and text title and text value 1`] = `
<table
  className="textAreaCriterion"
>
  <tbody>
    <tr>
      <td
        className="textAreaCriterionCell"
      >
        <div
          className="criterionTitle"
        >
          foo
        </div>
        <div>
          <textarea
            cols={80}
            onChange={[Function]}
            rows={4}
            value="bar"
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TextAreaView renders correctly when no selection and text title and value with XSS attack 1`] = `
<table
  className="textAreaCriterion"
>
  <tbody>
    <tr>
      <td
        className="textAreaCriterionCell"
      >
        <div
          className="criterionTitle"
        >
          title
        </div>
        <div>
          <textarea
            cols={80}
            onChange={[Function]}
            rows={4}
            value="<span onmouseover=\\"alert(document.cookie)\\">this is an
XSS attack</span>"
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TextAreaView renders correctly when no selection and text title and value with newlines 1`] = `
<table
  className="textAreaCriterion"
>
  <tbody>
    <tr>
      <td
        className="textAreaCriterionCell"
      >
        <div
          className="criterionTitle"
        >
          foo
        </div>
        <div>
          <textarea
            cols={80}
            onChange={[Function]}
            rows={4}
            value="bar

baz"
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TextAreaView renders correctly when selection and empty title and empty value 1`] = `
<table
  className="textAreaCriterion"
>
  <tbody>
    <tr>
      <td
        className="textAreaCriterionCell"
      >
        <div
          className="criterionTitle"
        />
        <div>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "",
              }
            }
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TextAreaView renders correctly when selection and text title and empty value 1`] = `
<table
  className="textAreaCriterion"
>
  <tbody>
    <tr>
      <td
        className="textAreaCriterionCell"
      >
        <div
          className="criterionTitle"
        >
          xyzzy
        </div>
        <div>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "",
              }
            }
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TextAreaView renders correctly when selection and text title and text value 1`] = `
<table
  className="textAreaCriterion"
>
  <tbody>
    <tr>
      <td
        className="textAreaCriterionCell"
      >
        <div
          className="criterionTitle"
        >
          foo
        </div>
        <div>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "bar",
              }
            }
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TextAreaView renders correctly when selection and text title and value with XSS attack 1`] = `
<table
  className="textAreaCriterion"
>
  <tbody>
    <tr>
      <td
        className="textAreaCriterionCell"
      >
        <div
          className="criterionTitle"
        >
          title
        </div>
        <div>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "&lt;span onmouseover=\\"alert(document.cookie)\\"&gt;this is an<br>XSS attack&lt;/span&gt;",
              }
            }
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TextAreaView renders correctly when selection and text title and value with newlines 1`] = `
<table
  className="textAreaCriterion"
>
  <tbody>
    <tr>
      <td
        className="textAreaCriterionCell"
      >
        <div
          className="criterionTitle"
        >
          foo
        </div>
        <div>
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "bar<br><br>baz",
              }
            }
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
